src/lib/JIT.hs

```diff
@@ -717,60 +657,67 @@ externDecl (ExternFunSpec fname retTy retAttrs funAttrs argTys) =
-      L.FloatingPointType L.DoubleFP -> dispatchOp fp64 ".f64" ""
-      L.FloatingPointType L.FloatFP  -> dispatchOp fp32 ".f32" "f"
-      _ -> error $ "Unsupported CPU floating point type: " ++ show (L.typeOf x)
-      where
-        dispatchOp ty llvmSuffix libmSuffix = case op of
-          Exp             -> callFloatIntrinsic ty $ "llvm.exp"   ++ llvmSuffix
-          Exp2            -> callFloatIntrinsic ty $ "llvm.exp2"  ++ llvmSuffix
-          Log             -> callFloatIntrinsic ty $ "llvm.log"   ++ llvmSuffix
-          Log2            -> callFloatIntrinsic ty $ "llvm.log2"  ++ llvmSuffix
-          Log10           -> callFloatIntrinsic ty $ "llvm.log10" ++ llvmSuffix
-          Log1p           -> callFloatIntrinsic ty $ "log1p"      ++ libmSuffix
-          Sin             -> callFloatIntrinsic ty $ "llvm.sin"   ++ llvmSuffix
-          Cos             -> callFloatIntrinsic ty $ "llvm.cos"   ++ llvmSuffix
-          Tan             -> callFloatIntrinsic ty $ "tan"        ++ libmSuffix
-          Sqrt            -> callFloatIntrinsic ty $ "llvm.sqrt"  ++ llvmSuffix
-          Floor           -> callFloatIntrinsic ty $ "llvm.floor" ++ llvmSuffix
-          Ceil            -> callFloatIntrinsic ty $ "llvm.ceil"  ++ llvmSuffix
-          Round           -> callFloatIntrinsic ty $ "llvm.round" ++ llvmSuffix
-          LGamma          -> callFloatIntrinsic ty $ "lgamma"     ++ libmSuffix
-          _ -> error $ "Unsupported CPU operation: " ++ show op
-        floatIntrinsic ty name = ExternFunSpec (L.mkName name) ty [] [] [ty]
-        callFloatIntrinsic ty name = emitExternCall (floatIntrinsic ty name) [x]
-
-    cpuBinaryIntrinsics op x y = case L.typeOf x of
-      L.FloatingPointType L.DoubleFP -> dispatchOp fp64 ".f64"
-      L.FloatingPointType L.FloatFP  -> dispatchOp fp32 ".f32"
-      _ -> error $ "Unsupported CPU floating point type: " ++ show (L.typeOf x)
-      where
-        dispatchOp ty llvmSuffix = case op of
-          FPow -> callFloatIntrinsic ty $ "llvm.pow" ++ llvmSuffix
-          _ -> error $ "Unsupported CPU operation: " ++ show op
-        floatIntrinsic ty name = ExternFunSpec (L.mkName name) ty [] [] [ty, ty]
-        callFloatIntrinsic ty name = emitExternCall (floatIntrinsic ty name) [x, y]
+    dispatchOp ty llvmSuffix = case op of
+      FPow -> callFloatIntrinsic ty $ "llvm.pow" ++ llvmSuffix
+      _ -> error $ "Unsupported CPU operation: " ++ show op
+    floatIntrinsic ty name = ExternFunSpec (L.mkName name) ty [] [] [ty, ty]
+    callFloatIntrinsic ty name = emitExternCall (floatIntrinsic ty name) [x, y]
+
+constStringFunction :: L.Name -> B.ByteString -> Function
+constStringFunction name str = runCompile CPU $ do
+  let chars = map (C.Int 8) $ map (fromIntegral . fromEnum) (B.unpack str) ++ [0]
+  ptr <- malloc i8 $ i64Lit $ length chars
+  let strVal = L.ConstantOperand $ C.Array i8 chars
+  tyPtr <- castLPtr (L.typeOf strVal) ptr
+  store tyPtr strVal
+  makeFunction name [] (Just ptr)
```

Comment:
The annoying thing was that that approach forced us to have two passes, because we need to actually call into LLVM to generate the kernel-as-string before building the LLVM code for all the non-CUDA stuff.

Reply:
I don't follow. What I proposed is only changing the way we embed the string literal (given as the byte string argument) in the llvm-hs AST. There's no calling into LLVM necessary. My helper did utilize an llvm-hs helper, but it's 100% pure and it was just simpler to do that than to construct the global definition myself.